truffleruby/src/main/java/org/truffleruby/core/thread/ThreadNodes.java

```diff
@@ -415,15 +415,9 @@ public DynamicObject wakeup(DynamicObject thread,
 
         @Specialization(guards = {"isRubyProc(unblocker)", "isRubyProc(runner)"})
         public Object unblock(VirtualFrame frame, DynamicObject thread, DynamicObject unblocker, DynamicObject runner) {
-            Layouts.THREAD.setUnblocker(thread, unblocker);
-            Layouts.THREAD.setStatus(thread, ThreadStatus.SLEEP);
-
-            try {
-                return yield(frame, runner);
-            } finally {
-                Layouts.THREAD.setUnblocker(thread, null);
-                Layouts.THREAD.setStatus(thread, ThreadStatus.RUN);
-            }
+            return getContext().getThreadManager().runUntilResult(this,
```

Comment:
I really don't know what the correct answer is here. I'll defer to you and @eregon. Since the `SafepointManager` can interrupt threads whenever it wants, not resuming the action could be incorrect as well.